pkg/services/baremetal/baremetal/baremetal.go

```diff
@@ -363,77 +366,121 @@ func (s *Service) getAssociatedHost(ctx context.Context) (*infrav1.HetznerBareMe
 	hosts := infrav1.HetznerBareMetalHostList{}
 	opts := &client.ListOptions{
 		Namespace: s.scope.BareMetalMachine.Namespace,
 	}
 
 	if err := s.scope.Client.List(ctx, &hosts, opts); err != nil {
-		return nil, nil, fmt.Errorf("failed to list hosts: %w", err)
+		return nil, nil, "", fmt.Errorf("failed to list hosts: %w", err)
 	}
 
 	labelSelector := s.getLabelSelector()
 
 	availableHosts := make([]*infrav1.HetznerBareMetalHost, 0, len(hosts.Items))
 
+	mapOfSkipReasons := make(map[string]int)
+
+	usableDevices := 0
 	for i, host := range hosts.Items {
 		if host.Spec.ConsumerRef != nil && consumerRefMatches(host.Spec.ConsumerRef, s.scope.BareMetalMachine) {
 			helper, err := patch.NewHelper(&hosts.Items[i], s.scope.Client)
 			if err != nil {
-				return nil, nil, fmt.Errorf("failed to create patch helper: %w", err)
-			}
-			return &hosts.Items[i], helper, nil
-		}
-		if s.scope.BareMetalMachine.Spec.InstallImage.Swraid == 1 {
-			// Machine should have RAID. Skip machines which have less than two WWNs
-			if len(host.Spec.RootDeviceHints.Raid.WWN) < 2 {
-				continue
+				return nil, nil, "", fmt.Errorf("failed to create patch helper: %w", err)
 			}
+			return &hosts.Items[i], helper, "", nil
 		}
 		if host.Spec.ConsumerRef != nil {
 			continue
 		}
+		if host.Spec.Status.ProvisioningState != infrav1.StateNone {
+			continue
+		}
+
+		// from now on each "continue" should add an entry
+		// to mapOfSkipReasons.
+		usableDevices++
+
+		if host.Spec.RootDeviceHints != nil &&
+			host.Spec.RootDeviceHints.IsValid() {
+			// If there are no RootDeviceHints, then the it should be ok that the hbmh gets picked.
+			// Then during registring, the HardwareDetails get fetched. After that the process fails.
+			// This has the benefit, that the user can easily copy+paste the WWNs from the HardwareDetails
+			// to the RootDeviceHints.
+			// Doing that without first choosing the hbmh would be nice, there is a feature request:
+			// https://github.com/syself/cluster-api-provider-hetzner/issues/1166
+			if s.scope.BareMetalMachine.Spec.InstallImage.Swraid == 1 {
+				// Machine should have RAID. Skip machines which have less than two WWNs
+				len_wwns := len(host.Spec.RootDeviceHints.Raid.WWN)
+				if len_wwns < 2 {
+					mapOfSkipReasons[fmt.Sprintf("machine-should-use-swraid-but-only-%d-RAID-WWN-in-hbmh", len_wwns)]++
+					continue
+				}
+			} else {
+				// Machine should have no RAID.
+				if host.Spec.RootDeviceHints.WWN == "" {
+					mapOfSkipReasons["machine-should-use-no-swraid-and-no-non-raid-WWN-in-hbmh"]++
+					continue
+				}
+			}
+		}
 		if host.Spec.MaintenanceMode != nil && *host.Spec.MaintenanceMode {
+			mapOfSkipReasons["hbmh-in-maintenance-mode"]++
 			continue
 		}
 		if host.GetDeletionTimestamp() != nil {
+			mapOfSkipReasons["hbmh-has-deletion-timestamp"]++
 			continue
 		}
 		if host.Spec.Status.ErrorMessage != "" {
+			mapOfSkipReasons["hbmh-has-error-message-in-status"]++
```

Comment:
put this after deletion timestamp

Reply:
@janiskemper please take a look at the new code. Does it make sense now? I moved "label" up, because I think we should not look too deep in to machines which are not in our scope.